\documentclass{article}
\usepackage[utf8]{inputenc}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{listings}
\usepackage{xcolor}
\newcommand{\code}[1]{`\texttt{#1}`}
\lstset{frame=single,backgroundcolor=\color{lightgray},numbers=left,breaklines=true,basicstyle=\ttfamily}


\title{TD Bioinfo}
\author{Pierre Lindenbaum - Institut du Thorax. Nantes. France}
\date{\today}

\begin{document}
\maketitle
\begin{abstract}
Here are few questions to prepare the courses \textbf{"Advanced NCBI"} and \textbf{"Next-Generation-Sequencing"}.\\
The aim of those exercices is to assure that you have a basic knowledge of the tools and vocabulary that will be used during the courses.\\
Only basic linux commands are required here, no external program should be installed.
\end{abstract}

\section{Preamble}
Before the course, you MUST \textbf{register an account  on \url{https://gitlab.univ-nantes.fr} } .

\section{Linux}
Briefly describe the usage of the following linux commands : man ;
cd ;
ls ;
pwd ;
ls ;
cp ;
rm ;
mv ;
cat ;
more ;
grep ;
sort ;
uniq ;
paste ;
join ;
tr ;
head ;
tail ;
mkdir ;
curl ;
awk ;
make ;
gzip ;
git\\

\noindent
What is 	the shell \textbf{completion ?} \\
What is 	the shell \textbf{PATH} ? \\
What is 	\textbf{stdout} ? \\
What is 	\textbf{stderr} ?\\
What is 	\textbf{stdin} ?\\
How do you 	\textbf{redirect} the ouput of a program to a file ?\\
How do you 	\textbf{append} the ouput of a program to a file ?\\
How do you 	write a \textbf{tabulation} in the shell command line ?\\
How would you diagnose/fix the following problem:

\begin{lstlisting}[language=bash]
$ blastn
blastn: command not found
\end{lstlisting}

In the HTTP protocol, what are the POST and GET methods ?\\
What is the JSON format ?\\
What is a XML well-formed document ?\\
What is a XML DTD ?\\
What is a XML Schema ?\\
What is a XML valid document ?\\
What is Xpath ?\\

%
% "Nicotiana tabacum"[ORGN] 100:150[SLEN] trnH[GENE] variation[FKEY]
% (Science[JOUR] AND COL1A1[GENE] AND Asara[AUTH]) NOT ("Mammut americanum" [ORGN] OR "Brachylophosaurus canadensis"[ORGN])

\section{General Bioinformatics}
\begin{itemize}
\item What is a 'SNP' ? 
\item What is a 'Genome build' ?
\item What is a 'Reference' Genome ?
\item What's the approximate size of the human genome ?
\item How many chromosomes is there in the human genome ?
\item What the approximate size of the longest human chromosome ?
\item What's the difference between a binary file and flat file ?
\item How many bytes do you need to store the \textbf{length} of the human genome ?
\item In a sequence containing only 'A', 'T', 'G' and 'C' , how many bases can you store in one byte ?
\item Cite some advantages/inconvenients of storing data in a tab delimited format vs using a structured format (e.g: XML, JSON, ASN.1)
\item Cite some advantages/inconvenients of handling data using a graphical interface (like Miscrosoft Excel) vs using the command line.
\item Find the following nucleotide entry in the NCBI:\emph{ a Sequence of "Nicotiana tabacum" for the gene 'trnH' with a length comprised between 100 and 150 nucleotides and having a feature "variation"}.
\item Find the following protein entry in the NCBI:\emph{ a sequence for the gene COL1A1 published by Dr. Asara in the "Science" journal,  that is not a sequence of "Mammut americanum" nor "Brachylophosaurus canadensis"}.
\end{itemize}

\section{Bash}
\noindent
What's the purpose of the following bash command line ? ( \url{https://gist.github.com/lindenb/65e98e5752ea26eb9868} )

\begin{lstlisting}[language=bash]
$ curl  "http://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/chrM.fa.gz" |\
gunzip -c |\
tail -n +2 |\
tr "[:lower:]" "[:upper:]" |\
tr -d '\n' |\
tr "AC" "TG" |\
sed 's/\(.\)/\1#/g' |\
tr "#" "\n" |\
LC_ALL=C sort |\
uniq -c |\
sed 's/^[ ]*//' |\
cut -d ' ' -f1 |\
tr "\n" " " |\
awk '{printf("%f %%\n",$1/($1+$2));}' > result.txt
\end{lstlisting}

\section{Makefile}
\noindent
(wikipedia:) \emph{"In software development, Make is a utility that automatically builds executable programs and libraries from source code by reading files called makefiles which specify how to derive the target program"}.\\You'll find many simple tutorials for \textbf{make} on the web.\\

Here is a file named \textbf{Makefile} ( \url{https://gist.github.com/lindenb/65e98e5752ea26eb9868} ).

\lstinputlisting[language=make,breaklines=true,numbers=left]{bioinfo2013.makefile}
Answer the following questions (shell should be \code{bash})
\begin{itemize}
\item type \code{make clean \&\& make}. What happens ?
\item what is a target ?
\item what is a dependency ?
\item what is the symbol  \code{\$@} ?
\item what is the symbol  \code{\$\textless} ?
\item what is the symbol  \code{\$\^} ?
\item where are the \code{tab (\textbackslash{}t)}  characters in the Makefile ?
\item what is the meaning of the line
\begin{lstlisting}[language=make]
%.rna:%.dna
	tr "Tt" "Uu" < $< > $@
\end{lstlisting}	
\item type \code{make clean \&\& make all}. What happens ?
\item why \code{all} was placed at the top ?
\item what is the default target ?
\item type \code{make clean \&\& make \&\& make \&\& make \&\& make \&\& make}. What happens ?
\item type \code{make clean \&\& make gamma.fa}. What happens ?
\item type \code{make clean \&\& make delta.fa}. What happens ?
\item type \code{make clean \&\& make \&\& make -B gamma.fa}. What happens ?
\item type \code{make clean \&\& make -n}. What happens ?
\item type \code{make clean \&\& make \&\& touch alpha.fa \&\& make }. What happens ?
\item type \code{make clean \&\& make \&\& rm alpha.fa \&\& make }. What happens ?
\item why \code{clean} and \code{all} were declared as \code{.PHONY}.
\item type \code{make clean \&\& make -j 3 all}. What happens ?
\item remove the line \code{.SECONDARY=} and type \code{make clean \&\& make all}. What happens ?
\item what is the benefit of using a makefile rather than a shell script ?
\end{itemize}

\section{The Human Genome}

The chromosomes for the latest Human build are available at: \url{http://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/}. Answer the following question using a linux command line:
\begin{itemize}
\item What's the length of the chr22 ?
\item Look at the 100000 first lines of the chr22. Explain what you see.
\item Get a count of each base in the chr22
\item Using the linux commands 'curl', 'gunzip', 'tr' and 'rev', get the reverse complement of the chromosome chrM.
\end{itemize}


\section{Using XSLT}
Many NCBI web-services produce a XML document. For example, the following URL is a XML-based list to the databases available from the  \textbf{NCBI}: \url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi}.\\
\begin{lstlisting}[language=xml]
<!DOCTYPE eInfoResult PUBLIC "-//NLM//DTD eInfoResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd">
<eInfoResult>
  <DbList>
    <DbName>pubmed</DbName>
    <DbName>protein</DbName>
    <DbName>nuccore</DbName>
    <DbName>nucleotide</DbName>
    <DbName>nucgss</DbName>
    <DbName>nucest</DbName>
    <DbName>structure</DbName>
    (...)
\end{lstlisting}
\textbf{XSLT} is a XML specification used to transform a XML to another type of document (XML, HTML or text).
As an example, the XSLT stylesheet \textbf{einfo2html.xsl} available at \url{https://gist.github.com/lindenb/65e98e5752ea26eb9868} transforms  \url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi} to a HTML document. The ouput would start with:\\
\begin{lstlisting}[language=html]
<html><body><ul>
<li><a>pubmed</a></li>
<li><a>protein</a></li>
<li><a>nuccore</a></li>
<li><a>nucleotide</a></li>
<li><a>nucgss</a></li>
<li><a>nucest</a></li>
<li><a>structure</a></li>
<li><a>genome</a></li>
<li><a>assembly</a></li>
    (...)
\end{lstlisting} 
\noindent
Using the command line XSLT processor \textbf{xsltproc}, generate this HTML and visualize it in a web browser.\\
The \textbf{href} attribute is missing in the \textless{}a/\textgreater  anchors. For example, the href for 'pubmed' would be :

\begin{lstlisting}[language=html]
<html><body><ul>
<li><a href="http://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=pubmed">pubmed</a></li>
    (...)
\end{lstlisting} 
\noindent
Modify the XSLT stylesheet: using the XSLT element \textless{}xsl:attribute/\textgreater  add the missing XML attributes. For example see: \url{http://stackoverflow.com/questions/3321119}.


\end{document}

